### PR TITLE
Remove DecodedType typealias (associatedtype) from Decodable protocol

### PR DESCRIPTION
**Pull request by ikesyo**
Addresses #97.

/cc @tarunon

### CODE REVIEW

**Comment by ikesyo** [comment]
The build failed on Linux with the following error:

``` bash
$ swift test
Compiling Swift Module 'Himotokitest' (6 sources)
/home/travis/build/ikesyo/Himotoki/Tests/Himotoki/DecodeErrorTest.swift:25:21: error: constructing an object of class type 'Self' with a metatype value must use a 'required' initializer
        return self.init(string: value)!
               ~~~~ ^
Foundation.NSURL:15:24: note: selected non-required initializer 'init(string:)'
    public convenience init?(string URLString: String)
                       ^
<unknown>:0: error: build had 1 command failures
error: exit(1): swift-build-tool -f /home/travis/build/ikesyo/Himotoki/.build/debug.yaml test
```

@tarunon Do you have any solution or workaround for this?

**Comment by tarunon** [comment]
Yes, `self.init(string: value)!` is `NSURL`, not a `Self`.
I think we should cast `NSURL` to `Self`, but we cannot use `Self` in implement.
I have one idea, I'll make new pr. :3

**Comment by ikesyo** [comment]
> Yes, `self.init(string: value)!` is `NSURL`, not a `Self`.
> I think we should cast `NSURL` to `Self`, but we cannot use `Self` in implement.

I'm not sure why this succeeds on OS X with SwiftPM, but fails on Linux. 🤔

**Comment by tarunon** [comment]
I don't know why... But I think Linux is correct.
Because if make NSURL subclass, cannot use `self.init(string: String)` without override.
It may be swift compiler bug.

**Comment by tarunon** [comment]
Auk, 

>  Yes, self.init(string: value)! is NSURL, not a Self.

I misunderstood this problem, sorry.
It is not necessary that `Self` has `init(string: String)` implement.

File: Sources/StandardLib.swift [hunk 1]
```diff
@@ -43,10 +43,12 @@ extension Bool: Decodable {
 }
 
 internal func castOrFail<T>(e: Extractor) throws -> T {
-    let rawValue = e.rawValue
+    return try castOrFail(e.rawValue)
+}
 
-    guard let result = rawValue as? T else {
-        throw typeMismatch("\(T.self)", actual: rawValue, keyPath: nil)
+internal func castOrFail<T>(any: Any) throws -> T {
```
Comment (tarunon) [inline Sources/StandardLib.swift]: I think this arguments type must be `Any?`.
And we need a test of `URLHolder` decode success case.

Comment (ikesyo) [inline Sources/StandardLib.swift]: > I think this arguments type must be `Any?`.

`Any` can take `Optional<Any>`, so it would be fine, I think.

**Comment by ikesyo** [comment]
Okay, this should be good to go now.

@tarunon Do you have further thoughts on this?

File: Sources/StandardLib.swift [hunk 2]
```diff
@@ -43,10 +43,12 @@ extension Bool: Decodable {
 }
 
 internal func castOrFail<T>(e: Extractor) throws -> T {
-    let rawValue = e.rawValue
+    return try castOrFail(e.rawValue)
+}
 
-    guard let result = rawValue as? T else {
-        throw typeMismatch("\(T.self)", actual: rawValue, keyPath: nil)
+internal func castOrFail<T>(any: Any?) throws -> T {
```
Comment (tarunon) [inline Sources/StandardLib.swift:49]: `func castOrFail<T>(any: Any?) throws -> T` is convenient function in `Decodable`, so I would like to it to public function, I think.
Or, adding `Decodable` extension is good for this.

Comment (ikesyo) [inline Sources/StandardLib.swift:49]: I'm not sure what the real world use case is. The users can implement `Decodable.decode(e: Extractor)` by using required initializer or also can use `Transformer` API for decoding non-final classes. Honestly I'm not willing to make it `public`.

Comment (tarunon) [inline Sources/StandardLib.swift:49]: I understand that now.
OK, I have no idea about this pr. Thanks @ikesyo san. :smile_cat:

Comment (ikesyo) [inline Sources/StandardLib.swift:49]: @tarunon I really appreciate your efforts on this! :+1: Many thanks :pray:

**Comment by ikesyo** [comment]
Merging this :ship:

**Pull request merged**